File: ios/PSMobile/Screens/SpecialValueEditors.swift
import SwiftUI
import PSMShared

/// Die Werte, die als Zeichenkette nicht zu bedienen sind.
///
/// PrusaSlicer haelt intern alles als Text, auch eine Bettform
/// ("0x0,250x0,250x210,0x210") und die Reinigungsmengen einer MMU (eine
/// Matrix mit bis zu 25 Zahlen). Im allgemeinen Renderer landen sie in
/// einem einzeiligen Feld - technisch richtig, praktisch unbrauchbar:
/// ein Tippfehler in der Mitte macht das Bett kaputt, und die Matrix
/// laesst sich so gar nicht lesen.
///
/// Gerechnet wird in `SpecialValueCodec` im gemeinsamen Modul. Hier
/// steht nur, wie es aussieht.
enum SpecialValueEditors {

    /// Welche Schluessel einen eigenen Bearbeiter haben.
    static func hasEditor(_ key: String) -> Bool {
        key == "bed_shape" || key == "wiping_volumes_matrix"
    }
}

/// Die Bettform.
///
/// Der haeufige Fall ist ein Rechteck - dafuer drei Zahlen statt einer
/// Punktliste. Wer eine andere Form hat (Delta, abgeschraegte Ecken),
/// sieht die Punkte einzeln und kann sie aendern; erfinden kann die
/// Oberflaeche sie nicht.
struct BedShapeEditor: View {

    @ObservedObject var model: SlicerModel
    var onClose: () -> Void

    @Environment(\.psScale) private var ps
    @State private var punkte: [SpecialValueCodec.BedPoint] = []
    @State private var breite = ""
    @State private var tiefe = ""

    var body: some View {
        VStack(alignment: .leading, spacing: ps.pt(14)) {
            Text(PsUiCatalog.tr("Bed shape"))
                .font(.system(size: ps.font(20)))
                .foregroundStyle(PrusaColors.textPrimary)

            if istRechteck {
                Text(st("Rectangular bed", "Rechteckiges Bett"))
                    .font(.system(size: ps.font(12)))
                    .foregroundStyle(PrusaColors.textMuted)
                HStack(spacing: ps.pt(10)) {
                    zahl(st("Width", "Breite"), $breite, kennung: "bett.breite")
                    zahl(st("Depth", "Tiefe"), $tiefe, kennung: "bett.tiefe")
                }
            } else {
                // Keine erfundene Vereinfachung: was nicht rechteckig ist,
                // wird als das gezeigt, was es ist.
                Text(st("Custom shape with \(punkte.count) points",
                        "Freie Form mit \(punkte.count) Punkten"))
                    .font(.system(size: ps.font(12)))
                    .foregroundStyle(PrusaColors.textMuted)
                ScrollView {
                    VStack(spacing: ps.pt(4)) {
                        ForEach(Array(punkte.enumerated()), id: \.offset) { i, punkt in
                            HStack(spacing: ps.pt(8)) {
                                Text("\(i + 1)")
                                    .font(.system(size: ps.font(11)))
                                    .foregroundStyle(PrusaColors.textMuted)
                                    .frame(width: ps.pt(24))
                                Text(String(format: "%.1f  ×  %.1f mm", punkt.x, punkt.y))
                                    .font(.system(size: ps.font(13)))
                                    .foregroundStyle(PrusaColors.textPrimary)
                                Spacer()
                            }
                        }
                    }
                }
                .frame(maxHeight: ps.pt(240))
            }

            if let flaeche = flaecheQcm {
                Text(st("Area", "Fläche") + String(format: ": %.0f cm²", flaeche))
                    .font(.system(size: ps.font(12)))
                    .foregroundStyle(PrusaColors.textMuted)
                    .accessibilityIdentifier("bett.flaeche")
            }

            HStack(spacing: ps.pt(12)) {
                Button(st("Cancel", "Abbrechen"), action: onClose)
                    .foregroundStyle(PrusaColors.textMuted)
                Spacer()
                if istRechteck {
                    Button {
                        uebernehmen()
                    } label: {
                        Text(st("Apply", "Übernehmen"))
                            .foregroundStyle(.white)
                            .padding(.horizontal, ps.pt(20))
                            .frame(height: ps.touch(48))
                            .background(PrusaColors.orange)
                            .clipShape(RoundedRectangle(cornerRadius: ps.pt(3)))
                    }
                    .buttonStyle(.plain)
                    .accessibilityIdentifier("bett.uebernehmen")
                }
            }
            Spacer()
        }
        .padding(ps.pt(20))
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        .background(PrusaColors.background)
        .overlay(alignment: .topLeading) { PSMarke(name: "bett") }
        .onAppear(perform: laden)
    }

    /// Vier Punkte, an den Achsen ausgerichtet, mit Ursprung im
    /// Nullpunkt - dann reichen zwei Zahlen.
    private var istRechteck: Bool {
        guard punkte.count == 4 else { return false }
        let xs = Set(punkte.map { round($0.x * 10) })
        let ys = Set(punkte.map { round($0.y * 10) })
        return xs.count == 2 && ys.count == 2
    }

    private var flaecheQcm: Double? {
        guard punkte.count >= 3 else { return nil }
        return SpecialValueCodec.shared.polygonArea(points: punkte) / 100
    }

    private func laden() {
        let roh = model.config("bed_shape") ?? ""
        punkte = SpecialValueCodec.shared.parseBedShape(value: roh) ?? []
        breite = String(format: "%.0f", (punkte.map(\.x).max() ?? 0) - (punkte.map(\.x).min() ?? 0))
        tiefe = String(format: "%.0f", (punkte.map(\.y).max() ?? 0) - (punkte.map(\.y).min() ?? 0))
    }

    private func uebernehmen() {
        guard let b = Double(breite.replacingOccurrences(of: ",", with: ".")),
              let t = Double(tiefe.replacingOccurrences(of: ",", with: ".")),
              b > 0, t > 0 else { return }
        let neu = [
            SpecialValueCodec.BedPoint(x: 0, y: 0),
            SpecialValueCodec.BedPoint(x: b, y: 0),
            SpecialValueCodec.BedPoint(x: b, y: t),
            SpecialValueCodec.BedPoint(x: 0, y: t),
        ]
        model.setConfig("bed_shape", SpecialValueCodec.shared.encodeBedShape(points: neu))
        onClose()
    }

    private func zahl(_ label: String,
                      _ text: Binding<String>,
                      kennung: String) -> some View {
        VStack(alignment: .leading, spacing: ps.pt(4)) {
            Text(label)
                .font(.system(size: ps.font(11)))
                .foregroundStyle(PrusaColors.textMuted)
            TextField("", text: text)
                .keyboardType(.numbersAndPunctuation)
                .font(.system(size: ps.font(14)))
                .foregroundStyle(PrusaColors.textPrimary)
                .padding(.horizontal, ps.pt(10))
                .frame(height: ps.touch(44))
                .background(PrusaColors.panelRaised)
                .clipShape(RoundedRectangle(cornerRadius: ps.pt(4)))
                .accessibilityIdentifier(kennung)
        }
    }

    private func st(_ english: String, _ german: String) -> String {
        SimpleModeState.shared.text(english: english, german: german)
    }
}

/// Die Reinigungsmengen einer MMU.
///
/// Eine Matrix: wie viel Material beim Wechsel von Position i nach j
/// verworfen wird. Als eine Zeile Zahlen ist sie nicht zu lesen - als
/// Gitter mit Zeilen- und Spaltenkoepfen schon.
struct WipingVolumesEditor: View {

    @ObservedObject var model: SlicerModel
    var onClose: () -> Void

    @Environment(\.psScale) private var ps
    @State private var werte: [Double] = []

    private var n: Int { Int(Double(werte.count).squareRoot().rounded()) }

    var body: some View {
        VStack(alignment: .leading, spacing: ps.pt(12)) {
            Text(PsUiCatalog.tr("Wipe tower"))
                .font(.system(size: ps.font(20)))
                .foregroundStyle(PrusaColors.textPrimary)
            Text(st("Millimetres purged when changing from the row to the column.",
                    "Millimeter, die beim Wechsel von der Zeile zur Spalte verworfen werden."))
                .font(.system(size: ps.font(12)))
                .foregroundStyle(PrusaColors.textMuted)
                .fixedSize(horizontal: false, vertical: true)

            if n < 2 {
                Text(st("Only one extruder - nothing to purge.",
                        "Nur ein Extruder - nichts zu verwerfen."))
                    .font(.system(size: ps.font(13)))
                    .foregroundStyle(PrusaColors.textMuted)
            } else {
                ScrollView([.horizontal, .vertical]) {
                    Grid(horizontalSpacing: ps.pt(4), verticalSpacing: ps.pt(4)) {
                        GridRow {
                            Text("").frame(width: ps.pt(28))
                            ForEach(0..<n, id: \.self) { spalte in
                                kopf("\(spalte + 1)")
                            }
                        }
                        ForEach(0..<n, id: \.self) { zeile in
                            GridRow {
                                kopf("\(zeile + 1)")
                                ForEach(0..<n, id: \.self) { spalte in
                                    zelle(zeile: zeile, spalte: spalte)
                                }
                            }
                        }
                    }
                    .padding(ps.pt(4))
                }
                .frame(maxHeight: ps.pt(320))
            }

            HStack {
                Button(st("Cancel", "Abbrechen"), action: onClose)
                    .foregroundStyle(PrusaColors.textMuted)
                Spacer()
                Button { uebernehmen() } label: {
                    Text(st("Apply", "Übernehmen"))
                        .foregroundStyle(.white)
                        .padding(.horizontal, ps.pt(20))
                        .frame(height: ps.touch(48))
                        .background(PrusaColors.orange)
                        .clipShape(RoundedRectangle(cornerRadius: ps.pt(3)))
                }
                .buttonStyle(.plain)
                .accessibilityIdentifier("reinigung.uebernehmen")
            }
            Spacer()
        }
        .padding(ps.pt(20))
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        .background(PrusaColors.background)
        .overlay(alignment: .topLeading) { PSMarke(name: "reinigung") }
        .onAppear(perform: laden)
    }

    private func kopf(_ text: String) -> some View {
        Text(text)
            .font(.system(size: ps.font(11)))
            .foregroundStyle(PrusaColors.textMuted)
            .frame(width: ps.pt(56), height: ps.pt(28))
    }

    /// Die Diagonale bleibt leer: von einer Position auf sich selbst
    /// wechselt niemand.
    @ViewBuilder private func zelle(zeile: Int, spalte: Int) -> some View {
        let index = zeile * n + spalte
        if zeile == spalte {
            Text("—")
                .font(.system(size: ps.font(12)))
                .foregroundStyle(PrusaColors.textMuted)
                .frame(width: ps.pt(56), height: ps.touch(40))
        } else {
            TextField("", text: Binding(
                get: { index < werte.count ? String(format: "%.0f", werte[index]) : "" },
                set: { neu in
                    if index < werte.count,
                       let z = Double(neu.replacingOccurrences(of: ",", with: ".")) {
                        werte[index] = z
                    }
                }))
                .keyboardType(.numberPad)
                .multilineTextAlignment(.center)
                .font(.system(size: ps.font(12)))
                .foregroundStyle(PrusaColors.textPrimary)
                .frame(width: ps.pt(56), height: ps.touch(40))
                .background(PrusaColors.panelRaised)
                .clipShape(RoundedRectangle(cornerRadius: ps.pt(3)))
                .accessibilityIdentifier("reinigung.\(zeile).\(spalte)")
        }
    }

    private func laden() {
        let roh = model.config("wiping_volumes_matrix") ?? ""
        werte = (SpecialValueCodec.shared.parseFloats(value: roh) ?? []).map { $0.doubleValue }
    }

    private func uebernehmen() {
        let text = SpecialValueCodec.shared.encodeFloats(
            values: werte.map { KotlinDouble(double: $0) })
        model.setConfig("wiping_volumes_matrix", text)
        onClose()
    }

    private func st(_ english: String, _ german: String) -> String {
        SimpleModeState.shared.text(english: english, german: german)
    }
}

File: android/app/src/main/java/de/psmobile/ui/SpecialValueEditors.kt
package de.psmobile.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import de.psmobile.ui.theme.PrusaColors
import de.psmobile.SlicerModel
import de.psmobile.shared.rules.SpecialValueCodec
import java.util.Locale
import kotlin.math.round
import kotlin.math.roundToInt
import kotlin.math.sqrt

/**
 * Die Werte, die als Zeichenkette nicht zu bedienen sind - Gegenstueck
 * zu `ios/PSMobile/Screens/SpecialValueEditors.swift`.
 *
 * PrusaSlicer haelt intern alles als Text, auch eine Bettform
 * ("0x0,250x0,250x210,0x210") und die Reinigungsmengen einer MMU.
 * Gerechnet wird in `SpecialValueCodec` im gemeinsamen Modul. Hier
 * steht nur, wie es aussieht.
 */
object SpecialValueEditors {

    /** Welche Schluessel einen eigenen Bearbeiter haben. */
    fun hasEditor(key: String): Boolean =
        key == "bed_shape" || key == "wiping_volumes_matrix"
}

/**
 * Die Bettform.
 *
 * Der haeufige Fall ist ein Rechteck - dafuer zwei Zahlen statt einer
 * Punktliste. Wer eine andere Form hat, sieht die Punkte einzeln;
 * erfinden kann die Oberflaeche sie nicht.
 */
@Composable
fun BedShapeEditor(model: SlicerModel, onClose: () -> Unit) {
    val ps = LocalPsScale.current
    var punkte by remember { mutableStateOf<List<SpecialValueCodec.BedPoint>>(emptyList()) }
    var breite by remember { mutableStateOf("") }
    var tiefe by remember { mutableStateOf("") }

    /// Vier Punkte, an den Achsen ausgerichtet - dann reichen zwei Zahlen.
    val istRechteck = punkte.size == 4 &&
        punkte.map { round(it.x * 10) }.toSet().size == 2 &&
        punkte.map { round(it.y * 10) }.toSet().size == 2

    val flaecheQcm: Double? =
        if (punkte.size >= 3) SpecialValueCodec.polygonArea(punkte) / 100 else null

    fun laden() {
        val roh = model.config("bed_shape") ?: ""
        punkte = SpecialValueCodec.parseBedShape(roh) ?: emptyList()
        breite = String.format(
            Locale.US, "%.0f",
            (punkte.maxOfOrNull { it.x } ?: 0.0) - (punkte.minOfOrNull { it.x } ?: 0.0),
        )
        tiefe = String.format(
            Locale.US, "%.0f",
            (punkte.maxOfOrNull { it.y } ?: 0.0) - (punkte.minOfOrNull { it.y } ?: 0.0),
        )
    }

    fun uebernehmen() {
        val b = breite.replace(",", ".").toDoubleOrNull() ?: return
        val t = tiefe.replace(",", ".").toDoubleOrNull() ?: return
        if (b <= 0 || t <= 0) return
        val neu = listOf(
            SpecialValueCodec.BedPoint(0.0, 0.0),
            SpecialValueCodec.BedPoint(b, 0.0),
            SpecialValueCodec.BedPoint(b, t),
            SpecialValueCodec.BedPoint(0.0, t),
        )
        model.setConfig("bed_shape", SpecialValueCodec.encodeBedShape(neu))
        onClose()
    }

    LaunchedEffect(Unit) { laden() }

    Box(
        Modifier
            .fillMaxSize()
            .background(PrusaColors.background),
        contentAlignment = Alignment.TopStart,
    ) {
        Column(
            Modifier
                .fillMaxSize()
                .padding(ps.pt(20)),
            verticalArrangement = Arrangement.spacedBy(ps.pt(14)),
            horizontalAlignment = Alignment.Start,
        ) {
            Text(PsUiCatalog.tr("Bed shape"), fontSize = ps.font(20), color = PrusaColors.textPrimary)

            if (istRechteck) {
                Text(st("Rectangular bed", "Rechteckiges Bett"), fontSize = ps.font(12), color = PrusaColors.textMuted)
                Row(horizontalArrangement = Arrangement.spacedBy(ps.pt(10))) {
                    Zahl(st("Width", "Breite"), breite, { breite = it }, kennung = "bett.breite")
                    Zahl(st("Depth", "Tiefe"), tiefe, { tiefe = it }, kennung = "bett.tiefe")
                }
            } else {
                // Keine erfundene Vereinfachung: was nicht rechteckig ist,
                // wird als das gezeigt, was es ist.
                Text(
                    st("Custom shape with ${punkte.size} points", "Freie Form mit ${punkte.size} Punkten"),
                    fontSize = ps.font(12),
                    color = PrusaColors.textMuted,
                )
                Column(
                    Modifier
                        .heightIn(max = ps.pt(240))
                        .verticalScroll(rememberScrollState()),
                    verticalArrangement = Arrangement.spacedBy(ps.pt(4)),
                ) {
                    punkte.forEachIndexed { i, punkt ->
                        Row(
                            Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.spacedBy(ps.pt(8)),
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            Text(
                                "${i + 1}",
                                fontSize = ps.font(11),
                                color = PrusaColors.textMuted,
                                modifier = Modifier.width(ps.pt(24)),
                            )
                            Text(
                                String.format(Locale.US, "%.1f  ×  %.1f mm", punkt.x, punkt.y),
                                fontSize = ps.font(13),
                                color = PrusaColors.textPrimary,
                            )
                            Spacer(Modifier.weight(1f))
                        }
                    }
                }
            }

            if (flaecheQcm != null) {
                Text(
                    st("Area", "Fläche") + String.format(Locale.US, ": %.0f cm²", flaecheQcm),
                    fontSize = ps.font(12),
                    color = PrusaColors.textMuted,
                    modifier = Modifier.testTag("bett.flaeche"),
                )
            }

            Row(
                Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(ps.pt(12)),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                TextButton(onClick = onClose) {
                    Text(st("Cancel", "Abbrechen"), color = PrusaColors.textMuted)
                }
                Spacer(Modifier.weight(1f))
                if (istRechteck) {
                    Box(
                        Modifier
                            .clip(RoundedCornerShape(ps.pt(3)))
                            .background(PrusaColors.orange)
                            .clickable { uebernehmen() }
                            .height(ps.touch(48))
                            .padding(horizontal = ps.pt(20))
                            .testTag("bett.uebernehmen"),
                        contentAlignment = Alignment.Center,
                    ) {
                        Text(st("Apply", "Übernehmen"), color = Color.White)
                    }
                }
            }
            Spacer(Modifier.weight(1f))
        }
        PSMarke(name = "bett")
    }
}

/** Ein beschriftetes Zahlenfeld - `zahl(_:_:kennung:)` im Original. */
@Composable
private fun Zahl(
    label: String,
    text: String,
    onTextChange: (String) -> Unit,
    kennung: String,
) {
    val ps = LocalPsScale.current
    Column(verticalArrangement = Arrangement.spacedBy(ps.pt(4)), horizontalAlignment = Alignment.Start) {
        Text(label, fontSize = ps.font(11), color = PrusaColors.textMuted)
        BasicTextField(
            value = text,
            onValueChange = onTextChange,
            singleLine = true,
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
            textStyle = TextStyle(fontSize = ps.font(14), color = PrusaColors.textPrimary),
            cursorBrush = SolidColor(PrusaColors.orange),
            modifier = Modifier
                .clip(RoundedCornerShape(ps.pt(4)))
                .background(PrusaColors.panelRaised)
                .height(ps.touch(44))
                .padding(horizontal = ps.pt(10))
                .testTag(kennung),
            decorationBox = { inner ->
                Box(Modifier.fillMaxSize(), contentAlignment = Alignment.CenterStart) { inner() }
            },
        )
    }
}

/**
 * Die Reinigungsmengen einer MMU - eine Matrix, wie viel Material beim
 * Wechsel von Position i nach j verworfen wird, als Gitter mit Zeilen-
 * und Spaltenkoepfen.
 */
@Composable
fun WipingVolumesEditor(model: SlicerModel, onClose: () -> Unit) {
    val ps = LocalPsScale.current
    var werte by remember { mutableStateOf<List<Double>>(emptyList()) }

    val n = sqrt(werte.size.toDouble()).roundToInt()

    fun laden() {
        val roh = model.config("wiping_volumes_matrix") ?: ""
        werte = SpecialValueCodec.parseFloats(roh) ?: emptyList()
    }

    fun uebernehmen() {
        val text = SpecialValueCodec.encodeFloats(werte)
        model.setConfig("wiping_volumes_matrix", text)
        onClose()
    }

    LaunchedEffect(Unit) { laden() }

    Box(
        Modifier
            .fillMaxSize()
            .background(PrusaColors.background),
        contentAlignment = Alignment.TopStart,
    ) {
        Column(
            Modifier
                .fillMaxSize()
                .padding(ps.pt(20)),
            verticalArrangement = Arrangement.spacedBy(ps.pt(12)),
            horizontalAlignment = Alignment.Start,
        ) {
            Text(PsUiCatalog.tr("Wipe tower"), fontSize = ps.font(20), color = PrusaColors.textPrimary)
            Text(
                st(
                    "Millimetres purged when changing from the row to the column.",
                    "Millimeter, die beim Wechsel von der Zeile zur Spalte verworfen werden.",
                ),
                fontSize = ps.font(12),
                color = PrusaColors.textMuted,
            )

            if (n < 2) {
                Text(
                    st("Only one extruder - nothing to purge.", "Nur ein Extruder - nichts zu verwerfen."),
                    fontSize = ps.font(13),
                    color = PrusaColors.textMuted,
                )
            } else {
                Column(
                    Modifier
                        .heightIn(max = ps.pt(320))
                        .horizontalScroll(rememberScrollState())
                        .verticalScroll(rememberScrollState())
                        .padding(ps.pt(4)),
                    verticalArrangement = Arrangement.spacedBy(ps.pt(4)),
                ) {
                    Row(horizontalArrangement = Arrangement.spacedBy(ps.pt(4)), verticalAlignment = Alignment.CenterVertically) {
                        Text("", modifier = Modifier.width(ps.pt(28)))
                        for (spalte in 0 until n) Kopf("${spalte + 1}")
                    }
                    for (zeile in 0 until n) {
                        Row(horizontalArrangement = Arrangement.spacedBy(ps.pt(4)), verticalAlignment = Alignment.CenterVertically) {
                            Kopf("${zeile + 1}")
                            for (spalte in 0 until n) {
                                Zelle(
                                    zeile = zeile, spalte = spalte, n = n, werte = werte,
                                    onWert = { index, z ->
                                        werte = werte.toMutableList().also { it[index] = z }
                                    },
                                )
                            }
                        }
                    }
                }
            }

            Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                TextButton(onClick = onClose) {
                    Text(st("Cancel", "Abbrechen"), color = PrusaColors.textMuted)
                }
                Spacer(Modifier.weight(1f))
                Box(
                    Modifier
                        .clip(RoundedCornerShape(ps.pt(3)))
                        .background(PrusaColors.orange)
                        .clickable { uebernehmen() }
                        .height(ps.touch(48))
                        .padding(horizontal = ps.pt(20))
                        .testTag("reinigung.uebernehmen"),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(st("Apply", "Übernehmen"), color = Color.White)
                }
            }
            Spacer(Modifier.weight(1f))
        }
        PSMarke(name = "reinigung")
    }
}

@Composable
private fun Kopf(text: String) {
    val ps = LocalPsScale.current
    Box(Modifier.size(ps.pt(56), ps.pt(28)), contentAlignment = Alignment.Center) {
        Text(text, fontSize = ps.font(11), color = PrusaColors.textMuted)
    }
}

/** Die Diagonale bleibt leer: von einer Position auf sich selbst wechselt niemand. */
@Composable
private fun Zelle(
    zeile: Int,
    spalte: Int,
    n: Int,
    werte: List<Double>,
    onWert: (Int, Double) -> Unit,
) {
    val ps = LocalPsScale.current
    val index = zeile * n + spalte
    if (zeile == spalte) {
        Box(Modifier.size(ps.pt(56), ps.touch(40)), contentAlignment = Alignment.Center) {
            Text("—", fontSize = ps.font(12), color = PrusaColors.textMuted)
        }
    } else {
        BasicTextField(
            value = if (index < werte.size) String.format(Locale.US, "%.0f", werte[index]) else "",
            onValueChange = { neu ->
                if (index < werte.size) {
                    neu.replace(",", ".").toDoubleOrNull()?.let { z -> onWert(index, z) }
                }
            },
            singleLine = true,
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
            textStyle = TextStyle(fontSize = ps.font(12), color = PrusaColors.textPrimary, textAlign = TextAlign.Center),
            cursorBrush = SolidColor(PrusaColors.orange),
            modifier = Modifier
                .clip(RoundedCornerShape(ps.pt(3)))
                .background(PrusaColors.panelRaised)
                .size(ps.pt(56), ps.touch(40))
                .testTag("reinigung.$zeile.$spalte"),
            decorationBox = { inner ->
                Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) { inner() }
            },
        )
    }
}
